### ISSUE TITLE
Rework Network Menu Options

### ISSUE THREAD

**Issue by Death259** [Death259/PiAssist#17]
The network menu should be redone to look and act like when a user connects to a Bluetooth device. This way the user doesn't have to type in the wifi SSID. There will need to be an option to allow the user to manually type in the SSID in case the SSID is not being broadcasted.

When connecting to a WiFi network, rather than have the network information stored in the /etc/network/interfaces file, there needs to be a wpa_supplicant that is referenced in order to allow for multiple SSIDs to be connectible at any given time.

**Comment by HerbFargus** [comment]
I'm also interested in this (hopefully to the point of integrating it into the retropie source code) as wicd-curses prevents cpu speed scaling- RetroPie needs something leaner like this by default.

**Comment by HerbFargus** [comment]
Just for reference in case you're interested. I modified your script on the section for network settings to include WEP and Open Networks- I tested it with my Pi and was able to get it to work with all 3. I agree with the above that it will probably be better with WPA supplicant but for the time being maybe my small changes will be useful to you.

https://github.com/HerbFargus/Raspberry-Pi-Wifi-Scripts/blob/master/raspberrypiwifi.sh

Thanks for your work on your script, it's great!

**Comment by Death259** [comment]
Fantastic! That's awesome work. I will try to take a look at your code and then see if I can do the wpa_supplicant portion. I've been working on the Dropbox backup and then the [RetroPie Subreddit](http://www.reddit.com/r/retropie).

**Comment by HerbFargus** [comment]
Ok, I've updated the script and added Hidden SSID's and Static IP's (for WPA) they can be easily adapted for WEP or Open Networks, but I figure WPA & WPA2 are the main ones people will be using. 

One note on Hidden SSID's- I think you have to use the HEX key rather than the pass key which can be generated with `wpa_passphrase "ssid" "passphrase"` and then it will generate something like this. so I used awk and sed to extract the hex key as a variable like you did with the essid list- that way people don't have to manually put in the hex key themselves. 

```
network={
        ssid="SSID"
        #psk="PASSPHRASE"
        psk=38497220976092fc2707a838e4d4385019256149f99f935be22c90159d3b8373
}
```

I pulled most of the information from here:
https://www.raspberrypi.org/forums/viewtopic.php?t=7592

For Hidden SSID's This covers basically all the details you'd need for WPA and WPA2:

```
wpa-driver wext
wpa-ssid UR_ESSID
#; wpa-ap-scan is 1 for visible and 2 for hidden SSID
wpa-ap-scan 2
#; wpa-proto is WPA for WPA1 (aka WPA) or RSN for WPA2
wpa-proto WPA 
#; wpa-pairwise and wpa-group is TKIP for WPA1 or CCMP for WPA2
wpa-pairwise TKIP
wpa-group TKIP
wpa-key-mgmt WPA-PSK
#; use "wpa_passphrase UR_ESSID UR_KEY" to generate UR_HEX_KEY
#; enter the result below
wpa-psk UR_HEX_KEY
```

As far as static IP's this script is just meant for wireless, I left the wired interface alone- so people will have to manually change their wired connection settings to get a static IP. Even though most of the details are pretty much standard by default I gave them inputboxes so people can choose their own just in case they are different (Like the gateway, netmask etc.)

Btw  I like what you've done with the reddit page.

I was thinking for future reference since my coding is pretty redundant it might be worth looking into simplifying the logic where there would be one option to connect to wifi like you have in your script now, and then it would open into a menu box with 4 security options  (WPA, WPA2, WEP, Open Network) then it would ask if they wanted to configure a static IP, then it would ask if it was a hidden ssid, then it would ask for the SSID, and then the SSID password, then it would create the interfaces file accordingly. I'm not entirely sure how wpa_supplicant handles different security types and all, but anyhow just a few thoughts.

**Comment by Death259** [comment]
Alrighty, I started looking at this some. I've changed the code to list WiFi networks to `iwlist wlan0 scan | grep ESSID | awk -F \" '{print $2}`. This cut out a lot of the text substitutions. I also changed it so that a menu of WiFi networks is given and then the user can just select one or if they choose "Other" then they will have to type in the SSID manually.

Edit: I hadn't even thought about open networks... or for that matter any other security other than WPA2 (since that's what I use). It looks like when we `iwlist wlan0 scan` we can see what security the network is using, so we may be able to use that to figure out the security and then go from there.

**Comment by HerbFargus** [comment]
Well people really should just be using WPA2, but not everyone has control over their router's configuration so they would need to be accommodated. I've seen quite a few people on the retropie forum with that predicament. 

That's brilliant, I'm quite fond of awk- it will be much more user friendly to integrate a radiolist of the ssids instead of needing to type them manually.

I had forgotten scanning also showed the security type- that would be really useful in automating things since a lot of users are unsure of their router security (mine actually does both WPA and wpa2)

**Comment by Death259** [comment]
I think I've found that the easiest way to identify an open network is to look at the "Encryption key:" for all secured networks in my area it is "On". For the one unsecure network it is "Off". WPA/WEP will take a little more thought. But I agree, they should definitely be supported.

**Comment by HerbFargus** [comment]
Would this be useful?http://stackoverflow.com/questions/17809912/parsing-iw-wlan0-scan-output

**Comment by Death259** [comment]
Great find!!! I think that will do almost exactly what we need.

Edit: Doesn't look like ReroPie has `iw`...
Edit2: May work with `iwlist` though... I'll have to look at this some more another day.

**Comment by HerbFargus** [comment]
I was looking at the encryption stuff a bit more and I think we could make it work with iwlist based off of this post:

http://ubuntuforums.org/showthread.php?t=1402284

I'll look into this a little later today- I want to test out your changes to the code and I may add one more option after the other choice for hidden ssids as they won't show up when scanned without extra tools installed. I'll try my hand at parsing the encryption types and if I can get satisfactory results with that I'll see what I can figure out with WPA_supplicant.

**Comment by Death259** [comment]
I don't think we will have to worry about the hidden SSIDs, as long as we allow them to manually enter the SSID, which we already do.

The other part about the encryption is fantastic. It looks like you have most of the groundwork lined out. Once we get the part to figure out what the security is, then we can compact the code.

**Comment by HerbFargus** [comment]
Ok- So I've tested it with everything and these are the results I've come up with- these should be just about every configuration you'll ever need for any type of network for `/etc/network/interfaces` and `/etc/wpa_supplicant/wpa_supplicant.conf`

# Interfaces (all options- we won't need these as we'll be using wpa_supplicant)

```
# First 6 lines are always the same (unless you want to configure a static IP)
auto lo

iface lo inet loopback
iface eth0 inet dhcp

allow-hotplug wlan0
auto wlan0
iface wlan0 inet dhcp

# WPA/WPA2
   wpa-ssid "NETWORK_NAME"
   wpa-psk "NETWORK_PASSWORD"

# WEP
  wireless-essid NETWORK_NAME
  wireless-key NETWORK_PASSWORD

#Open Network
  wireless-essid NETWORK_NAME
  wireless-mode managed

# Hidden SSID (Add the following line after the password)
  wpa-scan-ssid 1

#Static IP
iface wlan0 inet static #(formerly this line: "iface wlan0 inet dhcp")
address 192.168.0.110 #(This is the IP you want your raspberry pi to have)
netmask 255.255.255.0 #(This is almost always the same)
gateway 192.168.0.1 #(almost always the same as well. you can verify with netstat -nr)
```

# Interfaces when using wpa_supplicant

```
# First 5 Lines are basically always the same

auto lo

iface lo inet loopback
iface eth0 inet dhcp

allow-hotplug wlan0
auto wlan0

# wpa-roam (roaming PREFERRED)
iface wlan0 inet manual
wpa-roam /etc/wpa_supplicant/wpa_supplicant.conf
iface default inet dhcp

# wpa-conf (non roaming)
iface wlan0 inet dhcp
wpa-conf /etc/wpa_supplicant/wpa_supplicant.conf

# wpa-roam (roaming static IP PREFERRED)
iface wlan0 inet manual
wpa-roam /etc/wpa_supplicant/wpa_supplicant.conf
iface default inet static
address 192.168.0.110 #(This is the IP you want your raspberry pi to have)
netmask 255.255.255.0 #(This is almost always the same)
gateway 192.168.0.1 #(almost always the same as well. you can verify with netstat -nr)

# wpa-conf (non roaming static IP)
iface wlan0 inet static
wpa-conf /etc/wpa_supplicant/wpa_supplicant.conf
address 192.168.0.110 #(This is the IP you want your raspberry pi to have)
netmask 255.255.255.0 #(This is almost always the same)
gateway 192.168.0.1 #(almost always the same as well. you can verify with netstat -nr)

```

# wpa_supplicant.conf

```
# First two lines are basically always the same
ctrl_interface=DIR=/var/run/wpa_supplicant GROUP=netdev
update_config=1

# WPA/WPA2
network={
     ssid="NETWORK_NAME"
     psk="NETWORK_PASSWORD"
}

#WEP
network={
     ssid="NETWORK_NAME"
     key_mgmt=NONE
     wep_tx_keyidx=0 #this forces it to use wep_key0
     wep_key0=YOURWEPKEY
}

#Open Network
network={
    ssid="NETWORK_NAME"
    key_mgmt=NONE
}

#Hidden SSID (add after any encryption settings)
    scan_ssid=1
```

# Iwlist results by encryption type

Code:

```
sudo iwlist wlan0 scan | awk '/IE: WPA/ || /ESSID:/ || /IE: IEEE/ || /Encryption/'
```

Results

```
#WPA
Encryption key:on
ESSID:"NETWORK_NAME1"
IE:WPA Version 1

#WPA/WPA2
Encryption key:on
ESSID:"NETWORK_NAME2"
IE:WPA Version 1
IE: IEEE 802.11i/WPA2 Version 1

#WEP
Encryption key:on
ESSID:"NETWORK_NAME3"

#Open Network
Encryption key:off
ESSID:"NETWORK_NAME4"
```

**Comment by Death259** [comment]
Wow, excellent work! That's very comprehensive. I think with all that, we have enough to be able to dynamically generate the connection information based solely off of the user just selecting a SSID.

**Comment by HerbFargus** [comment]
The only thing that I don't think could be sorted with iwlist results is hidden ssids, but we could easily make it append the scan_ssid=1 anytime someone types their SSID in manually. Not sure how we would determine the encryption based on that though, unless we just assume anyone who wants to hide their ssid is likely to be using a stronger encryption than an open network or WEP. Or we could just append the scan_ssid=1 line to all the encryption configs since it still connects even if its not hidden.

**Comment by Death259** [comment]
With Windows, I think you have to enter in all of the information if your SSID is hidden, so we could maybe do that.

**Comment by HerbFargus** [comment]
That's probably a good idea. I figure the vast majority of people won't be using hidden SSID's anyhow.

**Comment by Death259** [comment]
I think I've gotten most if not all of the changes needed for this all to work, but I can't get any of the configurations using wpa_supplicant to work for the life of me. [HerbFargus](https://github.com/HerbFargus), do you have a working example of /etc/network/interfaces and the subsequent wpa_supplicant.conf? (don't forget to redact your password) Maybe just start with a wpa2 example...

Edit: Nevermind... I got wpa2 to work. Looks like I just need to change how I verify it works within the application.

**Comment by Death259** [comment]
Connecting to WiFi networks has been completely reworked as of commit f9774e21a9723a271f045d88414870208a9c1f27. This should allow for connecting to all types of WiFi networks (open, WEP, WPA/WPA2). The connection information gets written out to the `/etc/network/interfaces` and `/etc/wpa_supplicant/wpa_supplicant.conf` configuration files. The only issue with this I currently see is someone trying to add the same network twice. I've made that not allowed at this point, but I think in the future, I will make it so that it will update the settings for that specific WiFi network. I'll leave this issue open so that I can accomplish this.

Thanks for all of your help and research [HerbFargus](https://github.com/HerbFargus). You lined out all of the content needed for the two different configuration files and I was able to get things working without too much hassle.

Edit: Another issue is that it will erroneously state that it was unable to connect. I think this is caused by the application not waiting on the WiFi connection to occur, so then the application doesn't see an IP address.

**Comment by HerbFargus** [comment]
I apologise I was otherwise preoccupied,  It looks pretty good- a few more tweaks and it will be flawless.  I tested it out and its misidentifying my ecryption as WEP rather than WPA or WPA2.  and as far the message saying there was an issue connecting couldn't you just add a sleep perametre for as long as it takes until it connects? and then maybe just throw in a quick message box or something that says connecting to network or something like that

**Comment by Death259** [comment]
Doh! I thought I managed to fix all that. It may be a little before i can fix it, but i will look at it as soon as i can.Sleeping for X amount of time is a good idea, and I will probably do that.

**Comment by HerbFargus** [comment]
It's possible it may have to do with my router using both wpa and wpa2- the code may not know how to reconcile both.

**Comment by Death259** [comment]
Well, it checks to see if WPA is anywhere in the encryption used, so I'm sure it's something I messed up.

**Comment by HerbFargus** [comment]
When I was testing the retropie wifi module I realised that my network card puts out a lot more IE lines than others and it was giving a false WEP config even though it was giving WPA to others with less, so we figured out a way to resolve it.

Perhaps you can use this code as a reference for some ideas to make the code logic a little more strict:
https://github.com/RetroPie/RetroPie-Setup/blob/master/scriptmodules/supplementary/wifi.sh

The retropie wifi script is great and all but I like your script as it can be used as a standalone program without retropie.

**Comment by Death259** [comment]
I can definitely use that code. I see where I have already done a good chunk of it (just my own way) and won't have to redo everything. I think there's a lot of good ideas in there. I can already see what I need to do to fix the big issue you are experiencing, or so I believe. I won't get to work on it till this weekend, but I think I can hammer this out pretty quickly.

Thanks!

**Comment by Death259** [comment]
I went ahead and pretty much copied and pasted what joolswills had done, in order to provide a quicker fix. I wasn't able to narrow down where the logic problem was in my code. What joolswills has done is miles ahead of my work, so I'm going to take it and adapt it to more meet my needs. I've made a few modifications already with what just got published in commit 8e640cad205aa46f47965b7995ee488a3527ab90. Let me know if you still run into the issue with your network being misidentified.
